src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs

```diff
@@ -289,70 +287,78 @@ static void ThrowNodeIsNotSame(TrieNode node, TrieNode cachedNodeCopy) =>
-                    BlockCommitSet set = CurrentPackage;
-                    if (set is not null)
-                    {
-                        if (_logger.IsTrace) _logger.Trace($"Current root (block {blockNumber}): {root}, block {set.BlockNumber}");
-                        set.Seal(root);
-                    }
+                    throw new InvalidOperationException("Cannot start a new block commit when an existing one is still not closed");
+                }
 
-                    bool shouldPersistSnapshot = _persistenceStrategy.ShouldPersist(set.BlockNumber);
-                    if (shouldPersistSnapshot)
-                    {
-                        _currentBatch ??= _nodeStorage.StartWriteBatch();
-                        try
-                        {
-                            PersistBlockCommitSet(address, set, _currentBatch, writeFlags: writeFlags);
-                            PruneCurrentSet();
-                        }
-                        finally
-                        {
-                            // For safety we prefer to commit half of the batch rather than not commit at all.
-                            // Generally hanging nodes are not a problem in the DB but anything missing from the DB is.
-                            _currentBatch?.Dispose();
-                            _currentBatch = null;
-                        }
-                    }
-                    else
-                    {
-                        PruneCurrentSet();
-                    }
+                Monitor.Enter(_dirtyNodesLock);
+            }
 
-                    CurrentPackage = null;
-                    if (_pruningStrategy.PruningEnabled && Monitor.IsEntered(_dirtyNodesLock))
-                    {
-                        Monitor.Exit(_dirtyNodesLock);
-                    }
+            _currentBlockCommitter = new BlockCommitter(this, CreateCommitSet(blockNumber));
+            return _currentBlockCommitter;
+        }
+
+        private void FinishBlockCommit(BlockCommitSet set, TrieNode? root)
+        {
+            if (_logger.IsTrace) _logger.Trace($"Enqueued blocks {_commitSetQueue?.Count ?? 0}");
+            set.Seal(root);
+
+            bool shouldPersistSnapshot = _persistenceStrategy.ShouldPersist(set.BlockNumber);
+            if (shouldPersistSnapshot)
+            {
+                INodeStorage.WriteBatch currentBatch = _nodeStorage.StartWriteBatch();
+                try
+                {
+                    PersistBlockCommitSet(set, currentBatch);
+                    PruneCommitSet(set);
+                }
+                finally
+                {
+                    // For safety we prefer to commit half of the batch rather than not commit at all.
+                    // Generally hanging nodes are not a problem in the DB but anything missing from the DB is.
+                    currentBatch?.Dispose();
```

Comment:
```suggestion
                // For safety we prefer to commit half of the batch rather than not commit at all.
                // Generally hanging nodes are not a problem in the DB but anything missing from the DB is.
                using INodeStorage.WriteBatch currentBatch = _nodeStorage.StartWriteBatch();
                PersistBlockCommitSet(set, currentBatch);
                PruneCommitSet(set);
```